### PR TITLE
Minor refactoring to collections

### PR DESCRIPTION
**Pull request by ShreyasJejurkar**
1. Removed unwanted casting.
2. Simplified negation logic.

### CODE REVIEW

File: src/libraries/System.Collections/src/System/Collections/Generic/SortedDictionary.cs [hunk 1]
```diff
@@ -220,7 +220,10 @@ public void Add(TKey key, TValue value)
 
         public void Clear()
         {
-            _set.Clear();
+            if (_set.Count != 0)
```
Comment (stephentoub) [inline src/libraries/System.Collections/src/System/Collections/Generic/SortedDictionary.cs:223]: Why add this check?

File: src/libraries/System.Collections/src/System/Collections/Generic/SortedList.cs [hunk 2]
```diff
@@ -402,18 +402,21 @@ bool ICollection.IsSynchronized
...
         {
-            // clear does not change the capacity
-            version++;
-            // Don't need to doc this but we clear the elements so that the gc can reclaim the references.
-            if (RuntimeHelpers.IsReferenceOrContainsReferences<TKey>())
-            {
-                Array.Clear(keys, 0, _size);
-            }
-            if (RuntimeHelpers.IsReferenceOrContainsReferences<TValue>())
+            if (_size != 0)
```
Comment (stephentoub) [inline src/libraries/System.Collections/src/System/Collections/Generic/SortedList.cs:405]: Why add this check?

File: src/libraries/System.Collections/src/System/Collections/Generic/Stack.cs [hunk 3]
```diff
@@ -70,12 +70,15 @@ bool ICollection.IsSynchronized
         // Removes all Objects from the Stack.
         public void Clear()
         {
-            if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+            if (_size != 0)
```
Comment (stephentoub) [inline src/libraries/System.Collections/src/System/Collections/Generic/Stack.cs:73]: Same question.  All of these changes are adding a bit of code that penalizes the non-empty case to help the empty case.  Is the empty case common enough to warrant that?

File: src/libraries/System.Collections/src/System/Collections/Generic/SortedList.cs (hunk 2 above)
Comment (MCCshreyas) [reply src/libraries/System.Collections/src/System/Collections/Generic/SortedList.cs:405]: If the size is already 0, then there are no elements to clear from the internal list, so we can avoid those calls to `RuntimeHelpers` and can exit directly.

File: src/libraries/System.Collections/src/System/Collections/Generic/SortedDictionary.cs (hunk 1 above)
Comment (MCCshreyas) [reply src/libraries/System.Collections/src/System/Collections/Generic/SortedDictionary.cs:223]: If there are no elements then avoid the call to base Clear() method and just exit!

Comment (MCCshreyas) [reply src/libraries/System.Collections/src/System/Collections/Generic/SortedDictionary.cs:223]: Oh, if it's slow then I will remove this. My intention was that if the size is already 0, then calling clear won't make any sense, and just wanted to avoid that whole call to base Clear().

File: src/libraries/System.Collections/src/System/Collections/Generic/SortedList.cs (hunk 2 above)
Comment (huoyaoyuan) [reply src/libraries/System.Collections/src/System/Collections/Generic/SortedList.cs:405]: `RuntimeHelpers.IsReferenceOrContainsReferences` isn't a "call". It will be treated as a **constant** by JIT.

File: src/libraries/System.Collections/src/System/Collections/Generic/Stack.cs (hunk 3 above)
Comment (MCCshreyas) [reply src/libraries/System.Collections/src/System/Collections/Generic/Stack.cs]: Thanks, @stephentoub. I have reverted back those unwanted changes!

File: src/libraries/System.Collections/src/System/Collections/Generic/SortedList.cs [hunk 4]
```diff
@@ -404,7 +404,7 @@ public void Clear()
         {
             // clear does not change the capacity
             version++;
-            // Don't need to doc this but we clear the elements so that the gc can reclaim the references.
+            // Don't need to doc this but we clear the elements so that the gc can reclaim  the references.
```
Comment (eiriktsarpalis) [inline src/libraries/System.Collections/src/System/Collections/Generic/SortedList.cs:407]: You might want to revert this change.

File: src/libraries/System.Collections/src/System/Collections/Generic/Stack.cs [hunk 5]
```diff
@@ -72,7 +72,7 @@ public void Clear()
         {
             if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
             {
-                Array.Clear(_array, 0, _size); // Don't need to doc this but we clear the elements so that the gc can reclaim the references.
+                Array.Clear(_array, 0, _size); // Don't need to doc this but we clear the   elements so that the gc can reclaim the references.
```
Comment (eiriktsarpalis) [inline src/libraries/System.Collections/src/System/Collections/Generic/Stack.cs:75]: Same here.

File: src/libraries/System.Collections/src/System/Collections/Generic/SortedList.cs [hunk 6]
```diff
@@ -116,7 +116,7 @@ public SortedList(IComparer<TKey>? comparer)
         // Constructs a new sorted dictionary with a given IComparer
         // implementation and a given initial capacity. The sorted list is
         // initially empty, but will have room for the given number of elements
-        // before any reallocations are required. The elements of the sorted list
+        // before any re-allocations are required. The elements of the sorted list
```
Comment (eiriktsarpalis) [inline src/libraries/System.Collections/src/System/Collections/Generic/SortedList.cs:119]: What's wrong with "reallocation"?

File: src/libraries/System.Collections/src/System/Collections/Generic/SortedDictionary.cs [hunk 7]
```diff
@@ -353,7 +353,7 @@ ICollection IDictionary.Values
                     throw new ArgumentNullException(nameof(key));
                 }
 
-                if (value == null && !(default(TValue) == null))
```
Comment (eiriktsarpalis) [inline src/libraries/System.Collections/src/System/Collections/Generic/SortedDictionary.cs:356]: @stephentoub any reason to assume this form is intentional? Could the change break any intended optimization?

Comment (stephentoub) [reply src/libraries/System.Collections/src/System/Collections/Generic/SortedDictionary.cs:356]: They're indistinguishable:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgZgAIS6BhAHgBUA+Ogbyrv4b1gECLjoBZUgAoAlHQC8XACYwAZgEMArrgxS2cgITy6AO224A3HwG06w0ROKyFXA1JUbtu/QuNncuDJWlAC+QA===
I don't know why it was written this way, but both forms are equivalent.

**Review by eiriktsarpalis — APPROVED** [review]
LGTM. Thank you for your contribution!
